File: common/src/test/scala/com/taskadapter/connector/NewConfigSuggesterTest.scala
package com.taskadapter.connector

import com.taskadapter.connector.NewConfigSuggester.suggestedFieldMappingsForNewConfig
import com.taskadapter.connector.definition.FieldMapping
import com.taskadapter.model._
import org.junit.runner.RunWith
import org.scalatest.junit.JUnitRunner
import org.scalatest.{FunSpec, Matchers}

@RunWith(classOf[JUnitRunner])
class NewConfigSuggesterTest extends FunSpec with Matchers {

  private val summaryM = FieldMapping(Some(Summary), Some(Summary), true, null)
  private val descriptionM = FieldMapping(Some(Description), Some(Description), true, null)
  private val dueDateM = FieldMapping(Some(DueDate), Some(DueDate), true, null)

  it("suggests elements from left connector") {
    val list = suggestedFieldMappingsForNewConfig(
      Seq(Summary, Description, Assignee),
      Seq(Summary, Description))

    list should contain only(summaryM, descriptionM)
  }

  it("suggests elements from right connector") {
    val list = suggestedFieldMappingsForNewConfig(
      Seq(Summary, Description, DueDate),
      Seq(Summary, Description, Priority, DueDate))

    list should contain only(summaryM, descriptionM, dueDateM)
  }

  it("empty lists give empty result") {
    suggestedFieldMappingsForNewConfig(Seq(), Seq()) shouldBe empty
  }

  it("empty left list gives empty results") {
    suggestedFieldMappingsForNewConfig(Seq(), Seq(Summary)) shouldBe empty
  }

  it("empty right list gives empty results") {
    suggestedFieldMappingsForNewConfig(Seq(Summary), Seq()) shouldBe empty
  }

}

File: webshared/src/test/java/com/taskadapter/common/ui/NewConfigSuggesterTest.java
package com.taskadapter.common.ui;

import com.taskadapter.model.AllFields;
import com.taskadapter.model.Description$;
import com.taskadapter.model.DueDate$;
import org.junit.Test;

import java.util.Arrays;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

public class NewConfigSuggesterTest {
    private static final FieldMapping summaryM = new FieldMapping(AllFields.summary(),
            AllFields.summary(), true, null);
    private static final FieldMapping descriptionM = new FieldMapping(Description$.MODULE$,
            Description$.MODULE$, true, null);
    private static final FieldMapping dueDateM = new FieldMapping(DueDate$.MODULE$,
            DueDate$.MODULE$, true, null);

    @Test
    public void suggestsElementsFromLeftConnector() {
        var list = NewConfigSuggester.suggestedFieldMappingsForNewConfig(
                List.of(AllFields.summary(), AllFields.description(), AllFields.assigneeLoginName()),
                List.of(AllFields.summary(), AllFields.description()));

        assertThat(list).containsOnly(summaryM, descriptionM);
    }

    @Test
    public void suggestsElementsFromRightConnector() {
        var list = NewConfigSuggester.suggestedFieldMappingsForNewConfig(
                java.util.List.of(AllFields.summary(), AllFields.description(), AllFields.dueDate()),
                java.util.List.of(AllFields.summary(), AllFields.description(), AllFields.priority(), AllFields.dueDate()));

        assertThat(list).containsOnly(summaryM, descriptionM, dueDateM);
    }

    @Test
    public void emptyListsGiveEmptyResult() {
        assertThat(
                NewConfigSuggester.suggestedFieldMappingsForNewConfig(List.of(), List.of()))
                .isEmpty();
    }

    @Test
    public void emptyLeftListGivesEmptyResults() {
        assertThat(
                NewConfigSuggester.suggestedFieldMappingsForNewConfig(List.of(), Arrays.asList(AllFields.summary())))
                .isEmpty();
    }

    @Test
    public void emptyRightListGivesEmptyResults() {
        assertThat(
                NewConfigSuggester.suggestedFieldMappingsForNewConfig(Arrays.asList(AllFields.summary()), java.util.List.of()))
                .isEmpty();
    }

}
